elf_size_analyze/section.py

```diff
@@ -0,0 +1,174 @@
+        COMPRESSED = 'C'
+        UNKNOWN = 'x'
+        OS_SPECIFIC = 'o'
+        EXCLUDE = 'E'
+        PURECODE = 'y'
+        PROCESSOR_SPECIFIC = 'p'
+
+        @classmethod
+        def to_string(cls, flag):
+            for name, value in vars(cls).items():
+                if not name.startswith('_'):
+                    if value == flag:
+                        return name
+            return None
+
+    def __init__(self, **kwargs):
+        self.num = kwargs['num']
+        self.name = kwargs['name']
+        self.type = kwargs['type']
+        self.address = kwargs['address']
+        self.offset = kwargs['offset']
+        self.size = kwargs['size']
+        self.entry_size = kwargs['entry_size']
+        self.flags = kwargs['flags']
+        self.link = kwargs['link']
+        self.info = kwargs['info']
+        self.alignment = kwargs['alignment']
+
+    def is_writable(self):
+        return self.Flag.WRITE in self.flags
+
+    def occupies_memory(self):
+        # these are the only relevant sections for us
+        return self.Flag.ALLOC in self.flags
+
+    # these two methods are probably a big simplification
+    # as they may be true only for small embedded systems
+    def occupies_rom(self):
+        return self.occupies_memory() and \
+            self.type not in ['NOBITS']
+
+    def occupies_ram(self):
+        return self.occupies_memory() and self.is_writable()
+
+    @classmethod
+    def from_readelf_line(cls, line):
+        """
+        Create a Section from a line of `readelf -WS` output.
+        """
+        m = cls.pattern.match(line)
+        if not m:
+            log.debug('no match: ' + line.strip())
+            return None
+
+        # convert non-string values
+        m = m.groupdict()
+        m['num'] = int(m['num'])
+        m['address'] = int(m['address'], 16)
+        m['offset'] = int(m['offset'], 16)
+        m['size'] = int(m['size'], 16)
+        m['entry_size'] = int(m['entry_size'], 16)
+        # not sure if these are base-16 or base-10
+        m['link'] = int(m['link'], 10)
+        m['info'] = int(m['info'], 10)
+        m['alignment'] = int(m['alignment'], 10)
+
+        return Section(**m)
+
+    @classmethod
+    def print(cls, sections):
+        lines = []
+        for s in sections:
+            fields = [str(s.num), s.name, s.type,
+                      hex(s.address), sizeof_fmt(s.size),
+                      ','.join(cls.Flag.to_string(f) for f in s.flags)]
+            lines.append(fields)
+        sizes = [max(len(l[i]) for l in lines) for i in range(6)]
+        h_fmt = '{:%d}   {:%d}   {:%d}   {:%d}   {:%d}   {:%d}' % (*sizes, )
+        fmt = '{:>%d}   {:%d}   {:%d}   {:>%d}   {:>%d}   {:%d}' % (*sizes, )
+        header = h_fmt.format('N', 'Name', 'Type', 'Addr', 'Size', 'Flags')
+        separator = '=' * len(header)
+        top_header = '{:=^{size}s}'.format(' SECTIONS ', size=len(separator))
+        print(Color.BOLD + top_header  + Color.RESET)
```

Comment:
Missing import of `Color`, `python -m elf_size_analyze code.elf -P` results in error.